File: tests/src/commonTest/kotlin/primitives/PassPrimitive.kt
package primitives

import withLib
import kotlin.test.Test
import kotlin.test.assertTrue

class PassPrimitive {

    @Test
    fun passVoid() = withLib {
        assertTrue(natives.test.passVoid())
    }

    @Test
    fun passChar() = withLib {
        assertTrue(natives.test.passChar('a'))
    }

    @Test
    fun passBoolean() = withLib {
        assertTrue(natives.test.passBoolean(true))
    }

    @Test
    fun passByte() = withLib {
        assertTrue(natives.test.passByte(1.toByte()))
    }

    @Test
    fun passShort() = withLib {
        assertTrue(natives.test.passShort(1.toShort()))
    }

    @Test
    fun passInt() = withLib {
        assertTrue(natives.test.passInt(99))
    }

    @Test
    fun passLong() = withLib {
        assertTrue(natives.test.passLong(9223372036854775805L))
    }

    @Test
    fun passFloat() = withLib {
        assertTrue(natives.test.passFloat(99.9f))
    }

    @Test
    fun passDouble() = withLib {
        assertTrue(natives.test.passDouble(1.1))
    }

    @Test
    fun passString() = withLib {
        assertTrue(natives.test.passString("test string"))
    }
}

File: tests/natives/test/src/primitives/pass_primitive.c
#include <api.h>
#include <string.h>

// Consume

KBoolean passVoid() {
    return true;
}

KBoolean passChar(KChar arg) {
    return arg == 'a';
}

KBoolean passBoolean(KBoolean arg) {
    return arg == true;
}

KBoolean passByte(KByte arg) {
    return arg == 1;
}

KBoolean passShort(KShort arg) {
    return arg == 1;
}

KBoolean passInt(KInt arg) {
    return arg == 99;
}

KBoolean passLong(KLong arg) {
    return arg == 9223372036854775805;
}

KBoolean passFloat(KFloat arg) {
    return arg == 99.9f;
}

KBoolean passDouble(KDouble arg) {
    return arg == 1.1;
}

KBoolean passString(KString arg) {
    return arg.length == 11 && strncmp(arg.data, "test string", arg.length) == 0;
}

KBoolean passEnum(MyEnum arg) {
    return arg == MyEnum_CASE2;
}

KBoolean passStruct(MyStruct arg) {
    return arg.a == 1 &&
        arg.b == 2 &&
        arg.c == 3 &&
        arg.d == 4;
}